### PR TITLE
[marshal] Use MonoClass getters in marshal.c

### PR DESCRIPTION
**Pull request by lambdageek**
1. Add `m_class_offsetof_fieldName` functions
   They just call `MONO_STRUCT_OFFSET (MonoClass, fieldName)`.  If the `MonoClass` definition is hidden, you get prototypes from `mono/metadata/class-abi-details.h`, otherwise the functions are inlined in that header.
2. Add `mono_class_set_nonblittable` accessor function.
   I don't really understand why, but `mono_marshal_load_type_info` wants to overwrite the `MonoClass:blittable` flag sometimes.
3. Change everything in `marshal.c` to use `m_class_get_field` getters.  No functional change.

#6925

### CODE REVIEW

File: mono/metadata/class-accessors.c [hunk 1]
```diff
@@ -497,6 +497,11 @@ mono_class_set_failure (MonoClass *klass, MonoErrorBoxed *boxed_error)
 	return TRUE;
 }
 
+void
+mono_class_set_nonblittable (MonoClass *klass) {
```
Comment (vargaz) [inline mono/metadata/class-accessors.c]: This needs to be assume that the caller holds the loader lock.

**Comment by HinTak** [comment]
I am wondering if I can ask some question about this particular pull here . I am finally spending some time to get my own hand dirty try to get an old mixed-mode assembly application to work ( https://github.com/mono/mono/issues/10268 ) and is somewhat pleasantly surprised that with mono 5.12 or below, it asserts at the `g_assert` line after `mono_class_get_checked()` in `mono_marshal_set_callconv_from_modopt()` in `mono/metadata/marshal.c`. While with mono 5.14 it went a lot further (with `--verbose`) .

The only important difference within  `mono_marshal_set_callconv_from_modopt()` between those versions seems to be this pull.

So my question is, why would switching to getter/setter from direct access fixes problem with that particular assertion? Or, if this is not the fix for the assertion, what other changes around that time might be? I just like to understanding more about this pull.

**Comment by lambdageek** [comment]
@HinTak it's unlikely that switching from direct access to `MonoClass` fields to using getters had the effect you describe.

It's more likely a result of #7012 - I think part of what that PR did was to correct which `MonoImage` is associated with a custom modifier, which could explain why we don't fail immediately at `mono_class_get_checked`

**Comment by HinTak** [comment]
Thanks for taking the time to reply. I'll try hacking around those lines to see if I can learn more about how it all works (and why it failed and now it gotten a bit further).

If you have time, can you have a look at my hack to get the mixed mode app to work? ( https://github.com/mono/mono/issues/10268#issuecomment-423848912 ) and I'd appreciate any comments, e.g. if I am doing something obviously wrong. - before 5.12, it gaves a 300k log on --verbose, after 5.14, it gives a 500k one - and gotten further. With my hack, it seems to finish parsing metadata and gives a 1.5M log with --verbose, but P/Invoke seem to have completely broke :-).

**Comment by HinTak** [comment]
I check git log and it cannot be #7012 - #7012 entered master at the right time frame to master but wasn't cherry-pick/backported into 5.14, and it appears only at 5.16 onwards. Still, thanks a lot for the time, and I learned something just reading the #7012 issue entry and what follows.

**Pull request merged**